pkg/planner/core/rule_outer_to_inner_join.go

```diff
@@ -0,0 +1,155 @@
+func convertOuter2InnerJoins(p *LogicalJoin, predicates []expression.Expression) {
+	if p.JoinType != LeftOuterJoin && p.JoinType != RightOuterJoin && p.JoinType != InnerJoin {
+		return
+	}
+
+	innerTable := p.children[0]
+	outerTable := p.children[1]
+	if p.JoinType == LeftOuterJoin {
+		innerTable, outerTable = outerTable, innerTable
+	}
+
+	// first simplify embedded outer join.
+	if innerPlan, ok := innerTable.(*LogicalJoin); ok {
+		convertOuter2InnerJoins(innerPlan, predicates)
+	}
+	if outerPlan, ok := outerTable.(*LogicalJoin); ok {
+		convertOuter2InnerJoins(outerPlan, predicates)
+	}
+
+	if p.JoinType == InnerJoin {
+		return
+	}
+	// then simplify embedding outer join.
+	canBeSimplified := false
+	for _, expr := range predicates {
+		isOk := isNullFiltered(p.SCtx(), innerTable.Schema(), expr, outerTable.Schema())
+		if isOk {
+			canBeSimplified = true
+			break
+		}
+	}
+	if canBeSimplified {
+		p.JoinType = InnerJoin
+	}
+}
+
+func isNullFiltered(ctx base.PlanContext, innerSchema *expression.Schema, predicate expression.Expression, outerSchema *expression.Schema) bool {
+	// avoid the case where the predicate only refers to the schema of outerTable
+	if expression.ExprFromSchema(predicate, outerSchema) {
+		return false
+	}
+
+	switch expr := predicate.(type) {
+	case *expression.ScalarFunction:
+		if expr.FuncName.L == ast.LogicAnd {
+			if isNullFiltered(ctx, innerSchema, expr.GetArgs()[0], outerSchema) {
+				return true
+			} else {
+				return isNullFiltered(ctx, innerSchema, expr.GetArgs()[1], outerSchema)
+			}
+		} else if expr.FuncName.L == ast.LogicOr {
+			if !(isNullFiltered(ctx, innerSchema, expr.GetArgs()[0], outerSchema)) {
+				return false
+			} else {
+				return isNullFiltered(ctx, innerSchema, expr.GetArgs()[1], outerSchema)
+			}
+		} else {
+			return isNullFilteredOneExpr(ctx, innerSchema, expr)
+		}
+	default:
+		return isNullFilteredOneExpr(ctx, innerSchema, expr)
+	}
+}
+
+// isNullFilteredOneExpr check whether a condition is null-rejected
+// A condition would be null-rejected in one of following cases:
+// If it is a predicate containing a reference to an inner table that evaluates to UNKNOWN or FALSE when one of its arguments is NULL.
+// If it is a conjunction containing a null-rejected condition as a conjunct.
+// If it is a disjunction of null-rejected conditions.
+func isNullFilteredOneExpr(ctx base.PlanContext, schema *expression.Schema, expr expression.Expression) bool {
+	exprCtx := ctx.GetExprCtx()
+	expr = expression.PushDownNot(exprCtx, expr)
+	if expression.ContainOuterNot(expr) {
+		return false
+	}
+	sc := ctx.GetSessionVars().StmtCtx
+	if !exprCtx.IsInNullRejectCheck() {
+		exprCtx.SetInNullRejectCheck(true)
+		defer exprCtx.SetInNullRejectCheck(false)
+	}
+	for _, cond := range expression.SplitCNFItems(expr) {
+		result := expression.EvaluateExprWithNull(exprCtx, schema, cond)
+		x, ok := result.(*expression.Constant)
+		if !ok {
+			continue
+		}
+		if x.Value.IsNull() {
+			return true
```

Comment:
If that is a correct understanding, should we return true when this filter is evaluated as null?

Reply:
f(x) is null rejecting (evaluates to FALSE) when f(null) is FALSE or NULL/UNKNOWN. NULL is included since it is treated as FALSE. The logic is the same in original code.